src/gateway/routers/RouterAggregator.sol

```diff
@@ -0,0 +1,231 @@
+        } else if (!MessagesLib.isMessageProof(payload)) {
+            pendingMessages[messageHash] = payload;
+        }
+    }
+
+    /// @dev Governance on Centrifuge Chain can initiate message recovery. After the challenge period,
+    ///      the recovery can be executed. If a malign router initiates message recovery, governance on
+    ///      Centrifuge Chain can dispute and immediately cancel the recovery, using any other valid router.
+    ///
+    ///      Only 1 recovery can be outstanding per message hash. If multiple routers fail at the same time,
+    //       these will need to be recovered serially (increasing the challenge period for each failed router).
+    function _handleRecovery(bytes memory payload) internal {
+        if (MessagesLib.isInitiateMessageRecovery(payload)) {
+            (bytes32 messageHash, address router) = MessagesLib.parseInitiateMessageRecovery(payload);
+            recoveries[messageHash] = Recovery(block.timestamp + RECOVERY_CHALLENGE_PERIOD, router);
+            emit InitiateMessageRecovery(messageHash, router);
+        } else if (MessagesLib.isDisputeMessageRecovery(payload)) {
+            bytes32 messageHash = MessagesLib.parseDisputeMessageRecovery(payload);
+            return _disputeMessageRecovery(messageHash);
+        }
+    }
+
+    function disputeMessageRecovery(bytes32 messageHash) public auth {
+        _disputeMessageRecovery(messageHash);
+    }
+
+    function _disputeMessageRecovery(bytes32 messageHash) internal {
+        delete recoveries[messageHash];
+        emit DisputeMessageRecovery(messageHash);
+    }
+
+    function executeMessageRecovery(bytes calldata message) public {
+        bytes32 messageHash = keccak256(message);
+        Recovery storage recovery = recoveries[messageHash];
+        require(recovery.timestamp != 0, "RouterAggregator/message-recovery-not-initiated");
+        require(recovery.timestamp <= block.timestamp, "RouterAggregator/challenge-period-has-not-ended");
+
+        _handle(message, validRouters[recovery.router]);
+        delete recoveries[messageHash];
+    }
+
+    // --- Outgoing ---
+    /// @dev Sends 1 message to the first router with the full message, and n-1 messages to the other routers with
+    ///      proofs (hash of message). This ensures message uniqueness (can only be executed on the destination once).
+    function send(bytes calldata message) public {
+        require(msg.sender == address(gateway), "RouterAggregator/only-gateway-allowed-to-call");
+
+        uint256 numRouters = routers.length;
+        require(numRouters > 0, "RouterAggregator/not-initialized");
+
+        bytes memory proof = MessagesLib.formatMessageProof(message);
+        for (uint256 i; i < numRouters; ++i) {
+            RouterLike(routers[i]).send(i == PRIMARY_ROUTER_ID - 1 ? message : proof);
+        }
+
+        emit SendMessage(message);
+    }
+
+    // --- Helpers ---
+    function quorum() external view returns (uint8) {
```

Comment:
should we not pass the router id here?

Reply:
The quorum is stored per router, but actually global. This is just for removing another SLOAD on incoming messages. See https://github.com/centrifuge/liquidity-pools/pull/235/files/e5c72d8b2cb0c66a3633ce7119720b50147d9bc6#diff-74db641fd13d0d5c57a45b5507e2869b6a8c616d7c7f757d074e0a51d8843e90R40